include/treelite/typeinfo.h

```diff
@@ -0,0 +1,163 @@
+/*!
+ * Copyright (c) 2017-2020 by Contributors
+ * \file typeinfo.h
+ * \brief Defines TypeInfo class and utilities
+ * \author Hyunsu Cho
+ */
+
+#ifndef TREELITE_TYPEINFO_H_
+#define TREELITE_TYPEINFO_H_
+
+#include <cstdint>
+#include <typeinfo>
+#include <string>
+#include <unordered_map>
+#include <sstream>
+#include <utility>
+#include <type_traits>
+
+namespace treelite {
+
+/*! \brief Types used by thresholds and leaf outputs */
+enum class TypeInfo : uint8_t {
+  kInvalid = 0,
+  kUInt32 = 1,
+  kFloat32 = 2,
+  kFloat64 = 3
+};
+static_assert(std::is_same<std::underlying_type<TypeInfo>::type, uint8_t>::value,
+              "TypeInfo must use uint8_t as underlying type");
+
+/*! \brief conversion table from string to TypeInfo, defined in tables.cc */
+extern const std::unordered_map<std::string, TypeInfo> typeinfo_table;
+
+/*!
+ * \brief Get string representation of type info
+ * \param info a type info
+ * \return string representation
+ */
+inline std::string TypeInfoToString(treelite::TypeInfo type) {
+  switch (type) {
+  case treelite::TypeInfo::kInvalid:
+    return "invalid";
+  case treelite::TypeInfo::kUInt32:
+    return "uint32";
+  case treelite::TypeInfo::kFloat32:
+    return "float32";
+  case treelite::TypeInfo::kFloat64:
+    return "float64";
+  default:
+    throw std::runtime_error("Unrecognized type");
+    return "";
+  }
+}
+
+/*!
+ * \brief Convert a template type into a type info
+ * \tparam template type to be converted
+ * \return TypeInfo corresponding to the template type arg
+ */
+template <typename T>
+inline TypeInfo InferTypeInfoOf() {
+  if (std::is_same<T, uint32_t>::value) {
+    return TypeInfo::kUInt32;
+  } else if (std::is_same<T, float>::value) {
+    return TypeInfo::kFloat32;
+  } else if (std::is_same<T, double>::value) {
+    return TypeInfo::kFloat64;
+  } else {
+    throw std::runtime_error(std::string("Unrecognized Value type") + typeid(T).name());
+    return TypeInfo::kInvalid;
+  }
+}
+
+/*!
+ * \brief Given a TypeInfo, dispatch a function with the corresponding template arg. More precisely,
+ *        we shall call Dispatcher<T>::Dispatch() where the template arg T corresponds to the
+ *        `type` parameter.
+ * \tparam Dispatcher Function object that takes in one template arg.
+ *         It must have a Dispatch() static function.
+ * \tparam Parameter pack, to forward an arbitrary number of args to Dispatcher::Dispatch()
+ * \param type TypeInfo corresponding to the template arg T with which
+ *             Dispatcher<T>::Dispatch() is called.
+ * \param args Other extra parameters to pass to Dispatcher::Dispatch()
+ * \return Whatever that's returned by the dispatcher
+ */
+template <template<class> class Dispatcher, typename ...Args>
+inline auto DispatchWithTypeInfo(TypeInfo type, Args&& ...args) {
```

Comment:
It's hard to understand what the intended usage of `DispatchWithTypeInfo()` / `DispatchWithModelTypes()` is. Could you add an example (or reference a place in code) demonstrating it?

Reply:
Here is the idea. Let's say we are given a `TypeInfo` variable `type` at runtime. We want to call `Dispatcher<T>::Dispatch()` method where the template arg `T` corresponds to `type`, and `Dispatcher<T>` is a class that contains a custom function that we'd like to run.

For a concrete example, see https://github.com/dmlc/treelite/pull/198/files#diff-409f38fcc1a201a94e70922bbf7c44a6R777-R788 and https://github.com/dmlc/treelite/pull/196/files#diff-65416f46630c396d5db3df0ac92a0a02R130-R148.